### PR TITLE
Fix NDC workflow events reapplication issue

### PR DESCRIPTION
**Pull request by wxing1292**
* Allow NDC workflow events reapplication resetting current closed workflow.
* Workflow with no decision task started event still pending #2420

### CODE REVIEW

File: service/history/nDCTransactionMgr.go [hunk 1]
```diff
@@ -232,83 +244,119 @@ func (r *nDCTransactionMgrImpl) backfillWorkflow(
...
 
-	return targetWorkflow.getContext().updateWorkflowExecutionWithNew(
-		now,
-		mode,
-		nil,
-		nil,
-		transactionPolicy,
-		nil,
-	)
+	if err := targetWorkflow.getContext().reapplyEvents(
```
Comment (venkat1109) [inline service/history/nDCTransactionMgr.go]: reading through the code, its not obvious to me as to what we do here i.e. at this point, either targetWorkflow is not current workflow or this is not the active cluster. What do we now ? Can you add a comment to help the reader ?

File: service/history/nDCTransactionMgr.go [hunk 2]
```diff
@@ -232,83 +244,119 @@ func (r *nDCTransactionMgrImpl) backfillWorkflow(
...
-	// workflow events reapplication to self (self workflow being current workflow)
-	// we need to handle 2 cases
-	// 1. workflow still running -> just reapply
-	// 2. workflow closed -> TODO wait until https://github.com/uber/cadence/issues/2420
-	//  NOTE: also remember that workflow reset will acquire a lock on current workflow (this)
-
-	// simple case workflow still running (implies current workflow),
-	// no workflow reset necessary
-	if targetWorkflowActiveCluster == currentCluster {
-		// target workflow is active && target workflow is current workflow
```
Comment (venkat1109) [inline service/history/nDCTransactionMgr.go]: I think this comment is useful, unless it is no longer relevant, please don't delete this.

File: service/history/historyEngine.go [hunk 3]
```diff
@@ -2796,7 +2802,66 @@ func (e *historyEngineImpl) ReapplyEvents(
...
+				// need to reset target workflow (which is also the current workflow)
+				// to accept events to be reapplied
+				baseRunID := mutableState.GetExecutionInfo().RunID
+				resetRunID := uuid.New()
+				baseRebuildLastEventID := mutableState.GetPreviousStartedEventID()
+
+				// TODO when https://github.com/uber/cadence/issues/2420 is finished
+				//  remove this block
+				if baseRebuildLastEventID == common.EmptyEventID {
+					e.logger.Warn("cannot reapply event to a finished workflow",
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/historyEngine.go]: that is correct, how would you like to change the comments?

File: service/history/nDCTransactionMgr.go (hunk 1 above)
Comment (wxing1292) [reply service/history/nDCTransactionMgr.go]: sure

File: service/history/historyEngine.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (venkat1109) [reply service/history/historyEngine.go]: I just wanted to call out that its a finished workflow _which had no decisions scheduled at all_
One suggestion would be  "cannot reapply event to a finished workflow which had no decisions scheduled"

File: service/history/historyEngine.go [hunk 4]
```diff
@@ -2796,7 +2802,66 @@ func (e *historyEngineImpl) ReapplyEvents(
...
+						e.shard.GetDomainCache(),
+						e.shard.GetClusterMetadata(),
+						context,
+						mutableState,
+						noopReleaseFn,
+					),
+					eventsReapplicationResetWorkflowReason,
+					reapplyEvents,
+				); err != nil {
+					return nil, err
```
Comment (yux0) [inline service/history/historyEngine.go]: If the workflow cannot be reset due to no decision task, will the reapply retry forever?

Comment (wxing1292) [reply service/history/historyEngine.go]: line 2816 does the check and give up, until #2420 is finished

**Pull request merged**